source/server/config/stats/metrics_service.cc

```diff
@@ -0,0 +1,73 @@
+#include "server/config/stats/metrics_service.h"
+
+#include <string>
+
+#include "envoy/registry/registry.h"
+
+#include "common/config/well_known_names.h"
+#include "common/grpc/async_client_impl.h"
+#include "common/network/resolver_impl.h"
+#include "common/stats/grpc_metrics_service_impl.h"
+
+#include "api/metrics_service.pb.h"
+#include "api/metrics_service.pb.validate.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+class GrpcMetricsServiceClientFactoryImpl : public Stats::Metrics::GrpcMetricsServiceClientFactory {
+public:
+  GrpcMetricsServiceClientFactoryImpl(Upstream::ClusterManager& cluster_manager,
+                                      const std::string& cluster_name)
+      : cluster_manager_(cluster_manager), cluster_name_(cluster_name) {}
+
+  // Metrics::GrpcMetricsServiceClientPtr
+  Stats::Metrics::GrpcMetricsServiceClientPtr create() override {
+    return std::make_unique<Grpc::AsyncClientImpl<envoy::api::v2::StreamMetricsMessage,
+                                                  envoy::api::v2::StreamMetricsResponse>>(
+        cluster_manager_, cluster_name_);
+  };
+
+  Upstream::ClusterManager& cluster_manager_;
+  const std::string cluster_name_;
+};
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(grpc_metrics_streamer);
```

Comment:
In the case of the stat sink, you can kill all the singleton stuff. There is only a single sink in the server instantiated.